### ISSUE TITLE
MemoryStore publishing Committed out of data causal order

### ISSUE THREAD

**Issue by fnipo** [jet/equinox#265]
MemoryStore.fs is publishing Committed out of data causal order

[Commited publishing LOC](https://github.com/jet/equinox/blob/729146e5af90d6794a78ecc8c459eca195b38a8b/src/Equinox.MemoryStore/MemoryStore.fs#L27)

I did an workaround on a local version that mitigated the issue
By wrapping the streams ConcurrentDictionary with a lock, it enforces sequential operations on the dictionary and consequently sequential Commited publishings.

It also means a ConcurrentDictionary is not needed anymore

Below is the change done on this [LOC](https://github.com/jet/equinox/blob/729146e5af90d6794a78ecc8c459eca195b38a8b/src/Equinox.MemoryStore/MemoryStore.fs#L42)
```fsharp
let update () =
    let res = streams.AddOrUpdate(streamName, seedStream, updateValue) |> Written
    // raise here, once, as updateValue can conceptually be invoked multiple times
    committed.Trigger((FsCodec.StreamName.parse streamName, events))
    res
try lock streams update
```

**Comment by bartelink** [comment]
See also #264

**Comment by bartelink** [comment]
Released in 2.4.0 on `/v2`, also on `master`
